infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java

```diff
@@ -151,16 +163,164 @@ protected void internalStart(Map<String, String> startOptions) throws Infrastruc
       if (interrupted) {
         throw new InfrastructureException("Kubernetes environment start was interrupted");
       }
-      try {
-        throw e;
-      } catch (InfrastructureException rethrow) {
-        throw rethrow;
-      } catch (Exception wrap) {
-        throw new InternalInfrastructureException(e.getMessage(), wrap);
+      wrapAndRethrow(e);
+    }
+  }
+
+  /**
+   * Waits for readiness of given machines.
+   *
+   * @param machinesFutures machines futures to wait
+   * @param toCancelFutures futures that must be explicitly closed when any error occurs
+   * @param failure failure callback that is used to prevent subsequent steps when any error occurs
+   * @throws InfrastructureException when waiting for machines exceeds the timeout
+   * @throws InfrastructureException when any problem occurred while waiting
+   * @throws InterruptedException when the thread is interrupted while waiting machines
+   */
+  private void waitMachines(
+      List<CompletableFuture<Void>> machinesFutures,
+      List<CompletableFuture<?>> toCancelFutures,
+      CompletableFuture<Void> failure)
+      throws InfrastructureException, InterruptedException {
+    try {
+      final CompletableFuture<Void> allDone =
+          CompletableFuture.allOf(
+              machinesFutures.toArray(new CompletableFuture[machinesFutures.size()]));
+      CompletableFuture.anyOf(allDone, failure).get(machineStartTimeoutMin, TimeUnit.MINUTES);
```

Comment:
As far as I understand previously it was used for setting a timeout of one machine. Now it limits all machines. Please consider renaming configuration parameters to `che.infra.kubernetes.machines_start_timeout_min` to make that fact more clear. Does it make sense?